src/snowflake/connector/network.py

```diff
@@ -867,9 +869,25 @@ def _request_exec_wrapper(
                     exception=str(e),
                     stack_trace=traceback.format_exc(),
                 )
-            if no_retry:
-                return {}
             cause = e.args[0]
+            if no_retry:
+                logger.error(cause)
+                TelemetryService.get_instance().log_http_request_error(
+                    "HttpRequestError: %s" % str(cause),
+                    full_url,
+                    method,
+                    SQLSTATE_IO_ERROR,
+                    ER_FAILED_TO_REQUEST,
+                    retry_timeout=retry_ctx.total_timeout,
+                    retry_count=retry_ctx.cnt,
+                    exception=str(e),
+                    stack_trace=traceback.format_exc(),
+                )
+                if isinstance(cause, Error):
+                    Error.errorhandler_wrapper_from_cause(conn, cause)
+                else:
+                    self.handle_invalid_certificate_error(conn, full_url, cause)
+                return {}  # required for tests
```

Comment:
Since this code lives the exact same way just a couple of lines under this, maybe we should break them into a helper function to not have duplicate code